File: articles/traits-character-and-moral-evaluation/python/analyze_traits_character.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_traits_character_morality.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

TRAITS = ["honesty_humility", "conscientiousness", "agreeableness", "emotional_stability", "openness"]
LAYERS = ["moral_identity", "practical_judgment", "institutional_accountability", "power_pressure", "social_desirability_pressure"]
OUTCOMES = ["ethical_behavior", "integrity_rating", "trustworthiness_rating"]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"participant_id", "evaluation_context", *TRAITS, *LAYERS, *OUTCOMES}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

df["descriptive_trait_reliability"] = df[["honesty_humility", "conscientiousness", "agreeableness"]].mean(axis=1)
df["moral_character_index"] = df[OUTCOMES].mean(axis=1)
df["judgment_context_index"] = df[["practical_judgment", "institutional_accountability"]].mean(axis=1)
df["trait_character_gap"] = df["moral_character_index"] - df["descriptive_trait_reliability"]
df["trait_reliability_level"] = np.where(df["descriptive_trait_reliability"] > df["descriptive_trait_reliability"].median(), "higher_trait_reliability", "lower_trait_reliability")
df["character_level"] = np.where(df["moral_character_index"] > df["moral_character_index"].median(), "higher_character_evaluation", "lower_character_evaluation")
df["trait_character_profile"] = df["trait_reliability_level"] + "_" + df["character_level"]
df["gap_direction"] = np.where(df["trait_character_gap"] >= 0, "character_evaluation_exceeds_trait_reliability", "trait_reliability_exceeds_character_evaluation")

df.groupby("evaluation_context").agg(
    n=("participant_id", "count"),
    trait_reliability_mean=("descriptive_trait_reliability", "mean"),
    moral_character_mean=("moral_character_index", "mean"),
    moral_identity_mean=("moral_identity", "mean"),
    practical_judgment_mean=("practical_judgment", "mean"),
    accountability_mean=("institutional_accountability", "mean"),
    power_pressure_mean=("power_pressure", "mean"),
).reset_index().to_csv(OUT / "python_evaluation_context_summary.csv", index=False)

df.groupby("trait_character_profile").agg(
    n=("participant_id", "count"),
    trait_reliability_mean=("descriptive_trait_reliability", "mean"),
    moral_character_mean=("moral_character_index", "mean"),
    moral_identity_mean=("moral_identity", "mean"),
    practical_judgment_mean=("practical_judgment", "mean"),
    accountability_mean=("institutional_accountability", "mean"),
).reset_index().to_csv(OUT / "python_trait_character_profile_summary.csv", index=False)

df.groupby("gap_direction").agg(
    n=("participant_id", "count"),
    trait_reliability_mean=("descriptive_trait_reliability", "mean"),
    moral_character_mean=("moral_character_index", "mean"),
    practical_judgment_mean=("practical_judgment", "mean"),
    power_pressure_mean=("power_pressure", "mean"),
).reset_index().to_csv(OUT / "python_trait_character_gap_summary.csv", index=False)

corr_cols = TRAITS + LAYERS + OUTCOMES + ["descriptive_trait_reliability", "moral_character_index", "judgment_context_index", "trait_character_gap"]
df[corr_cols].corr().to_csv(OUT / "python_traits_character_correlations.csv")

models = {
    "ethical_behavior": ols(df, "ethical_behavior", TRAITS + ["moral_identity"]),
    "moral_character_index": ols(df, "moral_character_index", ["descriptive_trait_reliability", "moral_identity", "practical_judgment", "institutional_accountability"]),
    "trait_character_gap": ols(df, "trait_character_gap", ["moral_identity", "practical_judgment", "institutional_accountability", "power_pressure", "social_desirability_pressure"]),
    "note": "Synthetic demonstration only; not moral ranking, hiring guidance, legal evaluation, or social sorting."
}
(OUT / "python_traits_character_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_traits_character_morality.csv", index=False)
print(f"Wrote outputs to {OUT}")

File: articles/traits-character-and-moral-evaluation/r/analyze_traits_character.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_traits_character_morality.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

traits <- c("honesty_humility", "conscientiousness", "agreeableness", "emotional_stability", "openness")
outcomes <- c("ethical_behavior", "integrity_rating", "trustworthiness_rating")

data$descriptive_trait_reliability <- rowMeans(data[, c("honesty_humility", "conscientiousness", "agreeableness")])
data$moral_character_index <- rowMeans(data[, outcomes])
data$judgment_context_index <- rowMeans(data[, c("practical_judgment", "institutional_accountability")])
data$trait_character_gap <- data$moral_character_index - data$descriptive_trait_reliability
data$trait_character_profile <- paste(
  ifelse(data$descriptive_trait_reliability > median(data$descriptive_trait_reliability), "higher_trait_reliability", "lower_trait_reliability"),
  ifelse(data$moral_character_index > median(data$moral_character_index), "higher_character_evaluation", "lower_character_evaluation"),
  sep = "_"
)

context_summary <- aggregate(
  data[, c("descriptive_trait_reliability", "moral_character_index", "moral_identity", "practical_judgment", "institutional_accountability", "power_pressure")],
  by = list(evaluation_context = data$evaluation_context),
  FUN = mean
)
write.csv(context_summary, file.path(outputs, "r_evaluation_context_summary.csv"), row.names = FALSE)

profile_summary <- aggregate(
  data[, c("descriptive_trait_reliability", "moral_character_index", "moral_identity", "practical_judgment", "institutional_accountability")],
  by = list(trait_character_profile = data$trait_character_profile),
  FUN = mean
)
write.csv(profile_summary, file.path(outputs, "r_trait_character_profile_summary.csv"), row.names = FALSE)

corr_cols <- c(traits, "moral_identity", "practical_judgment", "institutional_accountability", "power_pressure", "social_desirability_pressure", outcomes, "descriptive_trait_reliability", "moral_character_index", "judgment_context_index", "trait_character_gap")
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_traits_character_correlations.csv"))

model_ethical <- lm(ethical_behavior ~ honesty_humility + conscientiousness + agreeableness + emotional_stability + openness + moral_identity, data = data)
model_character <- lm(moral_character_index ~ descriptive_trait_reliability + moral_identity + practical_judgment + institutional_accountability, data = data)
model_gap <- lm(trait_character_gap ~ moral_identity + practical_judgment + institutional_accountability + power_pressure + social_desirability_pressure, data = data)

capture.output(summary(model_ethical), file = file.path(outputs, "r_ethical_behavior_model_summary.txt"))
capture.output(summary(model_character), file = file.path(outputs, "r_moral_character_model_summary.txt"))
capture.output(summary(model_gap), file = file.path(outputs, "r_trait_character_gap_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_traits_character_morality.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")
